### PR TITLE
Unit Tests for KubernetesClusterSchedulerBackend

### PR DESCRIPTION
**Pull request by mccheah**
Requires #454, which in turn requires #452.

### CODE REVIEW

**Comment by ifilonenko** [comment]
Thank you for this 👍

**Comment by mccheah** [comment]
@varunkatta I changed some things in the scheduler backend related to #244. Some of them are style things and variable renames, but there was a case I identified where we didn't call `removeExecutor` when I think we should. The unit tests reflect what I thought should be the expected behavior. It would be much appreciated if you could take a look at the changes and the tests to verify whether or not we're doing the right thing here.

**Comment by mccheah** [comment]
Integration test failure is legit from #454.

**Comment by mccheah** [comment]
This is ready for review, but is not strictly complete. I'm certain that some corner cases are missing in the scheduler backend tests. As well, we don't yet unit test the components that were factored out in #454 and #452. I'll probably revisit these but I would like to keep these PRs small so that we can incrementally follow the cases that we've been covering in these tests.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala [hunk 1]
```diff
@@ -29,25 +29,28 @@ import scala.collection.{concurrent, mutable}
...
-    kubernetesClient: KubernetesClient)
-  extends CoarseGrainedSchedulerBackend(scheduler, sc.env.rpcEnv) {
+    kubernetesClient: KubernetesClient,
+    allocatorExecutor: ScheduledExecutorService,
+    requestExecutorsService: ExecutorService)
+  extends CoarseGrainedSchedulerBackend(scheduler, rpcEnv) {
 
   import KubernetesClusterSchedulerBackend._
 
+  private val EXECUTOR_ID_COUNTER = new AtomicLong(0L)
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala:53]: moved this here so there's less static state?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala [hunk 2]
```diff
@@ -143,43 +138,47 @@ private[spark] class KubernetesClusterSchedulerBackend(
...
+      if (reasonCheckCount >= MAX_EXECUTOR_LOST_REASON_CHECKS) {
+        removeExecutor(executorId, SlaveLost("Executor lost for unknown reasons."))
+        deleteExecutorFromApiAndDataStructures(executorId)
       } else {
-        executorReasonChecks.put(executorId, reasonCheckCount + 1)
+        executorReasonCheckAttemptCounts.put(executorId, reasonCheckCount + 1)
+      }
+    }
+
+    def deleteExecutorFromApiAndDataStructures(executorId: String): Unit = {
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala]: `deleteExecutorFromClusterAndDataStructures`

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala [hunk 3]
```diff
@@ -411,17 +413,24 @@ private[spark] class KubernetesClusterSchedulerBackend(
...
         }
-        failedPods.put(pod.getMetadata.getName, exitReason)
+        podsWithKnownExitReasons.put(pod.getMetadata.getName, exitReason)
     }
 
     def handleDeletedPod(pod: Pod): Unit = {
-      val exitReason = ExecutorExited(getExecutorExitStatus(pod), exitCausedByApp = false,
-        "Pod " + pod.getMetadata.getName + " deleted or lost.")
-        failedPods.put(pod.getMetadata.getName, exitReason)
+      val alreadyReleased = isPodAlreadyReleased(pod)
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala]: inline `alreadyReleased`

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala [hunk 4]
```diff
@@ -433,12 +442,15 @@ private[spark] class KubernetesClusterSchedulerBackend(
...
     extends DriverEndpoint(rpcEnv, sparkProperties) {
-    private val externalShufflePort = conf.getInt("spark.shuffle.service.port", 7337)
 
     override def onDisconnected(rpcAddress: RpcAddress): Unit = {
       addressToExecutorId.get(rpcAddress).foreach { executorId =>
         if (disableExecutor(executorId)) {
-            executorsToRemove.add(executorId)
+          RUNNING_EXECUTOR_PODS_LOCK.synchronized {
+            runningExecutorsToPods.get(executorId).foreach { pod =>
+              disconnectedPodsByExecutorIdPendingRemoval(executorId) = pod
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala:448]: can you explain a bit why it's safe to remove the executor directly here, rather than going through the `executorsToRemove` set first?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala (hunk 1 above)
Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala:53]: Correct - it's much more difficult to unit test if the counter is global, because between different tests one needs to know what the counter is set to.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala (hunk 4 above)
Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala:448]: We don't remove the executor directly here. Some of the logic has changed and variables are renamed, so `executorsToRemove` doesn't exactly exist anymore anyways.

Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala:448]: Here we are marking the executor as disconnected and the allocator thread will clean it up once the exit reason is known. I believe this is the same semantics as before.

**Comment by mccheah** [comment]
Assigning @varunkatta and @foxish to verify correctness of the nuanced but important logic changes done here.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala (hunk 4 above)
Comment (varunkatta) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala:448]: Yup. No change in semantics just that the marking is done after verifying that the executor is still running/tracked.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala [hunk 5]
```diff
@@ -411,17 +412,23 @@ private[spark] class KubernetesClusterSchedulerBackend(
...
         }
-        failedPods.put(pod.getMetadata.getName, exitReason)
+        podsWithKnownExitReasons.put(pod.getMetadata.getName, exitReason)
     }
 
     def handleDeletedPod(pod: Pod): Unit = {
-      val exitReason = ExecutorExited(getExecutorExitStatus(pod), exitCausedByApp = false,
-        "Pod " + pod.getMetadata.getName + " deleted or lost.")
-        failedPods.put(pod.getMetadata.getName, exitReason)
+      val exitMessage = if (isPodAlreadyReleased(pod)) {
```
Comment (varunkatta) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala:423]: Thanks for this change..Makes it consistent with the case of ErroredPod handling

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala [hunk 6]
```diff
@@ -143,43 +138,47 @@ private[spark] class KubernetesClusterSchedulerBackend(
...
     def removeExecutorOrIncrementLossReasonCheckCount(executorId: String): Unit = {
-      val reasonCheckCount = executorReasonChecks.getOrElse(executorId, 0)
-      if (reasonCheckCount > MAX_EXECUTOR_LOST_REASON_CHECKS) {
-        removeExecutor(executorId, SlaveLost("Executor lost for unknown reasons"))
-        executorsToRecover.add(executorId)
-        executorReasonChecks -= executorId
+      val reasonCheckCount = executorReasonCheckAttemptCounts.getOrElse(executorId, 0)
+      if (reasonCheckCount >= MAX_EXECUTOR_LOST_REASON_CHECKS) {
+        removeExecutor(executorId, SlaveLost("Executor lost for unknown reasons."))
+        deleteExecutorFromClusterAndDataStructures(executorId)
```
Comment (varunkatta) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackend.scala:167]: +1 for the change. Thanks for catching the omission here.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackendSuite.scala [hunk 7]
```diff
@@ -0,0 +1,383 @@
+/*
```
Comment (varunkatta) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterSchedulerBackendSuite.scala:1]: Huge thanks for this. This unit test one of most important additions to keep the cluster backend healthy, correct and maintainable.

**Review by varunkatta — APPROVED** [review]
Changes look good. Makes the code more readable. Most importantly this change fixes a subtle bug, and adds unit-tests which is a huge win! Thanks Matt!

**Comment by varunkatta** [comment]
Also, I think this should probably go in before #392 . I can merge changes from this PR into #392 once this PR is accepted.

**Comment by mccheah** [comment]
Rebased

**Comment by mccheah** [comment]
rerun unit tests please

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterManager.scala [hunk 8]
```diff
@@ -132,12 +132,18 @@ private[spark] class KubernetesClusterManager extends ExternalClusterManager wit
         executorInitContainerBootstrap,
         executorInitContainerSecretVolumePlugin,
         kubernetesShuffleManager)
+    val allocatorExecutor = ThreadUtils
+        .newDaemonSingleThreadScheduledExecutor("kubernetes-pod-allocator")
+    val requestExecutorsService = ThreadUtils.newDaemonCachedThreadPool(
+        "kubernetes-request-executors")
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterManager.scala]: did you mean to change this name from `kubernetes-executor-requests` ?

**Review by ash211 — APPROVED** [review]
One minor nit that might not be worth a change, but otherwise really happy to see tests coming in to this class!

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterManager.scala (hunk 8 above)
Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/kubernetes/KubernetesClusterManager.scala]: Not particularly.

**Comment by ash211** [comment]
K reverted it -- will merge when builds are green!

**Comment by duyanghao** [comment]
@ash211 When will this be accepted?

**Comment by ash211** [comment]
It's merged -- was there more you thought should be done here?

**Comment by duyanghao** [comment]
@ash211 Sorry,i have just seen it,will it be closed soon?
i just think There may be some unknown problems in #244 as it still lacks large-scale use.
Anyway,i will test this function in practical use and report problems.

**Comment by ash211** [comment]
I've seen some problems myself with executor recovery, but haven't dug into why yet.  Please do open new issues with any observations you see of bad behavior!
